bot.py

```diff
@@ -1,59 +1,50 @@
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+"""Launches the Discord bot."""
+
 from discord.ext import commands
 
-import botcore.perms
-from botcore.config import config
-from botcore.verify import Verify
-from botcore.sign import Sign
-
-# Set up Firebase connection
-cred = credentials.Certificate("config/firebase_credentials.json")
-default_app = firebase_admin.initialize_app(cred)
-db = firestore.client()
-print("Logged in to Firebase")
-
-# Get verification secret
-try:
-    secret = db.collection("secrets").document("verify").get().to_dict()["secret"]
-except TypeError:
-    print("Generating new verification secret...")
-    secret = secrets.token_bytes(64)
-    db.collection("secrets").document("verify").set({"secret": secret})
-    print("Saved verification secret in Firebase")
-else:
-    print("Fetched verification secret from Firebase")
-
-# Set up mail server
-mail = smtplib.SMTP(host=config["smtp-server"], port=config["smtp-port"])
-mail.starttls()
-mail.login(config["email-address"], config["email-password"])
-print("Logged in to mail server")
-
-bot = commands.Bot(command_prefix=config["command-prefix"])
-
-@bot.event
-async def on_ready():
-    print(f"Bot running with command prefix {bot.command_prefix}")
-
-@bot.event
-async def on_command_error(ctx, error):
-    await ctx.send(str(error))
-
-@bot.command(name="exit")
-@botcore.perms.is_admin_user()
-async def cmd_exit(ctx):
-    await ctx.send("I am shutting down...")
-    mail.quit()
-    await bot.logout()
-    print("Successfully logged out. Exiting...")
-
-bot.add_cog(Verify(bot, secret, db, mail))
-bot.add_cog(Sign(bot))
-
-bot.run(config["bot-token"])
+import iam.perms
+from iam.config import config
+from iam.core import Core
+from iam.db import Database
+from iam.mail import Mail
+from iam.verify import Verify
+from iam.sign import Sign
+
+def main():
+    bot = commands.Bot(command_prefix=config["command-prefix"])
+
+    bot.load_extension("iam.core")
```

Comment:
Why switch to extensions?

Reply:
Extensions allow for hot reloading of code, as opposed to directly loading the cogs. May be useful in the future for updating the bot. Though this feature will likely not be used, it doesn't cost anything to have it